### DISCUSSION TITLE
Advice needed on validation with MapRequestPayload

### DISCUSSION THREAD

**Discussion by yard-mschwartz** [symfony/symfony#53492] [Q&A] [answered]
Formerly, when I still used the `Sensio/FrameworkExtraBundle`, denormalization of JSON-requests into objects worked slightly differently in comparison to the relatively new `MapRequestPayload` in Symfony's core.

This is an example of an older project, using `ParamConverter` in `Sensio/FrameworkExtraBundle` (also featuring `FriendsOfSymfony/RestBundle`):

```
#[Rest\Put('/api/{version}/seeds/{seed}/settings', name: 'seed_update_settings')]
#[ParamConverter('updateSeedSettings', converter: 'fos_rest.request_body')]
public function updateSettings(UpdateSeedSettings $updateSeedSettings, Seed $seed): View
{
    $updateSeedSettings->setSeedId($seed->getId());

    $envelope = $this->commandBus->dispatch($updateSeedSettings);

    // more code here
}
```

And this is an example from a project I'm working on right now:

```
#[Route('/api/change_password', name: 'api_change_password', methods: ["POST"], format: 'json')]
public function changePassword(
    #[MapRequestPayload(acceptFormat: 'json')] ChangePassword $changePassword,
): JsonResponse {
    $this->commandBus->dispatch($changePassword);

    return $this->json([]);
}
```

Note that in both examples, we are denormalizing the request body into a DTO, that is also a command, processed by a command handler, using Symfony's message system.

Ignoring the differences in syntax, the main difference here is that the first example doesn't run validation for the command until the command is being dispatched. The second example runs validation twice. Not only when dispatching the command but also when denormalizing the request body.

This change in behavior is challenging, because often the command needs to be supplemented before validation makes sense. In the first example you see that the seed id, coming from the url, needs to be added to the DTO before I dispatch the command.

What would be the best way to go about this? Simply prevent the validation step triggered by using `MapRequestPayload`? If so what would be the best way to do this? Abusing `validationGroups` does the trick, but that doesn't feel right.

**Comment by MatTheCat** [comment, marked as answer]
In your first example `$updateSeedSettings` is actually invalid until you set its `$seedId`, but this is only because you consider the request’s payload as a command — which is wrong. The payload actually contains seed settings which is an object you can validate standalone. Then, you can instantiate a command given the settings object and the seed’s ID. The command needs to be validated as well because the settings may already be validated in some contexts, but you cannot guarantee it.

Two validations are needed because an invalid payload is legit but not an invalid command, so you generally don’t want to handle violations the same way.

That’s why it also makes sense to validate `$changePassword` twice, even if I agree it is useless in your example.

**Comment by yard-mschwartz** [comment]
Bit of a late response by me, although I had read your reply earlier. I came to the same conclusion. We're just going to prevent validation initially, and defer that until the command is being dispatched. I rather not write boilerplate code that would be needed if you have different objects for the request body and the command that is being dispatched.
